DOI.js

```diff
@@ -35,23 +35,58 @@
 	***** END LICENSE BLOCK *****
 */
 
-// builds a list of DOIs
-function getDOIs(doc) {
-	// TODO Detect DOIs more correctly.
-	// The actual rules for DOIs are very lax-- but we're more strict.
-	// Specifically, we should allow space characters, and all Unicode
-	// characters except for control characters. Here, we're cheating
-	// by not allowing ampersands, to fix an issue with getting DOIs
-	// out of URLs.
-	// Additionally, all content inside <noscript> is picked up as text()
-	// by the xpath, which we don't necessarily want to exclude, but
-	// that means that we can get DOIs inside node attributes and we should
-	// exclude quotes in this case.
-	// DOI should never end with a period or a comma (we hope)
-	// Description at: http://www.doi.org/handbook_2000/appendix_1.html#A1-4
-	const DOIre = /\b10\.[0-9]{4,}\/[^\s&"']*[^\s&"'.,]/g;
-
-	var dois = [];
+// TODO Detect DOIs more correctly.
+// The actual rules for DOIs are very lax-- but we're more strict.
+// Specifically, we should allow space characters, and all Unicode
+// characters except for control characters. Here, we're cheating
+// by not allowing ampersands, to fix an issue with getting DOIs
+// out of URLs.
+// Additionally, all content inside <noscript> is picked up as text()
+// by the xpath, which we don't necessarily want to exclude, but
+// that means that we can get DOIs inside node attributes and we should
+// exclude quotes in this case.
+// DOI should never end with a period or a comma (we hope)
+// Description at: http://www.doi.org/handbook_2000/appendix_1.html#A1-4
+const DOIre = /\b10\.[0-9]{4,}\/[^\s&"']*[^\s&"'.,]/g;
+
+/**
+ * @return {string | string[]} A single string if the URL contains a DOI
+ * 		and the document contains no others, or an array of DOIs otherwise
+ */
+function getDOIs(doc, url) {
+	let fromURL = getDOIsFromURL(url);
+	let fromDocument = getDOIsFromDocument(doc);
+	if (
+		// We got a single DOI from the URL
+		fromURL.length == 1 && (
+			// And none from the document
+			fromDocument.length == 0
+			// Or one from the document, but the same one that was in the URL
+			|| fromDocument.length == 1 && fromDocument[0] == fromURL[0]
+		)
+	) {
+		return fromURL;
+	}
+	// De-duplicate before returning
+	return Array.from(new Set([...fromURL, ...fromDocument]));
+}
+
+function getDOIsFromURL(url) {
+	// Split on # and ?, so that we don't allow DOIs to contain those characters
+	// but do allow finding DOIs on either side of them (e.g. a DOI in the URL hash)
+	let urlParts = url.split(/[#?]/);
+	for (let urlPart of urlParts) {
+		let match = DOIre.exec(urlPart);
+		if (match) {
+			// Only return a single DOI from the URL
+			return [match[0]];
```

Comment:
If we're only returning a single DOI from the URL anyway, why not make this `getDOIFromURL()`, return a single DOI string or `false`/`null`, and not bother with the extra `.length == 1` and `[0]` and `...` in `getDOIs()`?

Reply:
I guess because then you would have to test if `fromURL` was null and not put it in the Set at the end.… Still, a little confusing to call it `getDOIsFromURL()`, so maybe better to just deal with that in `getDOIs()`.